### PR TITLE
Proper multi-gpu support with PPO

### PR DESCRIPTION
**Pull request by vwxyzjn**
## Description
This is a follow up on https://github.com/vwxyzjn/cleanrl/pull/162

## Types of changes

- [x] New algorithm

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [x] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [x] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [x] I have contacted @vwxyzjn to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [x] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [x] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [x] I have explained note-worthy implementation details.
    - [x] I have explained the logged metrics.
    - [x] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [x] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [x] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [x] I have added links to the tracked experiments.
- [ ] I have updated the tests accordingly (if applicable).

### CODE REVIEW

File: cleanrl/ppo_atari_multigpu.py [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+    return args
+
+
+def make_env(env_id, seed, idx, capture_video, run_name):
+    def thunk():
+        env = gym.make(env_id)
+        env = gym.wrappers.RecordEpisodeStatistics(env)
+        if capture_video:
+            if idx == 0:
+                env = gym.wrappers.RecordVideo(env, f"videos/{run_name}")
```
Comment (yooceii) [inline cleanrl/ppo_atari_multigpu.py:101]:
```
 if capture_video and idx == 0:
            env = gym.wrappers.RecordVideo(env, f"videos/{run_name}")
```

File: cleanrl/ppo_atari_multigpu.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+if __name__ == "__main__":
+    # torchrun --standalone --nnodes=1 --nproc_per_node=2 ppo_atari_multigpu.py
+    # taken from https://pytorch.org/docs/stable/elastic/run.html
+    local_rank = int(os.getenv("LOCAL_RANK", "0"))
+    world_size = int(os.getenv("WORLD_SIZE", "1"))
+    args = parse_args()
+    args.world_size = world_size
+    args.num_envs = int(args.num_envs / world_size)
+    args.batch_size = int(args.num_envs * args.num_steps)
+    args.minibatch_size = int(args.batch_size // args.num_minibatches)
```
Comment (yooceii) [inline cleanrl/ppo_atari_multigpu.py:163]: Dup to line 89-90?

File: cleanrl/ppo_atari_multigpu.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+            "hyperparameters",
+            "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+        )
+
+    # TRY NOT TO MODIFY: seeding
+    # CRUCIAL: note that we needed to pass a different seed for each data parallelism worker
+    args.seed += local_rank
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed - local_rank)
```
Comment (yooceii) [inline cleanrl/ppo_atari_multigpu.py:202]: Why not
```suggestion 
 torch.manual_seed(args.seed)
 args.seed += local_rank
 random.seed(args.seed)
 np.random.seed(args.seed)
```

File: cleanrl/ppo_atari_multigpu.py [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+            device = torch.device(f"cuda:{local_rank}" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = gym.vector.SyncVectorEnv(
+        [make_env(args.env_id, args.seed + i, i, args.capture_video, run_name) for i in range(args.num_envs)]
+    )
+    assert isinstance(envs.single_action_space, gym.spaces.Discrete), "only discrete action space is supported"
+
+    agent = Agent(envs).to(device)
+    torch.manual_seed(args.seed)
```
Comment (yooceii) [inline cleanrl/ppo_atari_multigpu.py:222]: Dup to line 201?

File: cleanrl/ppo_atari_multigpu.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+    actions = torch.zeros((args.num_steps, args.num_envs) + envs.single_action_space.shape).to(device)
+    logprobs = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    rewards = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    dones = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    values = torch.zeros((args.num_steps, args.num_envs)).to(device)
+
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    start_time = time.time()
+    next_obs = torch.Tensor(envs.reset()).to(device)
```
Comment (yooceii) [inline cleanrl/ppo_atari_multigpu.py:236]:
```suggestion
 next_obs = torch.tensor(envs.reset()).to(device)
```
per https://discuss.pytorch.org/t/difference-between-torch-tensor-and-torch-tensor/30786/2

File: cleanrl/ppo_atari_multigpu.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+            with torch.no_grad():
+                action, logprob, _, value = agent.get_action_and_value(next_obs)
+                values[step] = value.flatten()
+            actions[step] = action
+            logprobs[step] = logprob
+
+            # TRY NOT TO MODIFY: execute the game and log data.
+            next_obs, reward, done, info = envs.step(action.cpu().numpy())
+            rewards[step] = torch.tensor(reward).to(device).view(-1)
+            next_obs, next_done = torch.Tensor(next_obs).to(device), torch.Tensor(done).to(device)
```
Comment (yooceii) [inline cleanrl/ppo_atari_multigpu.py:262]: Same to line 236

File: cleanrl/ppo_atari_multigpu.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+                nn.utils.clip_grad_norm_(agent.parameters(), args.max_grad_norm)
+                optimizer.step()
+
+            if args.target_kl is not None:
+                if approx_kl > args.target_kl:
+                    break
+
+        y_pred, y_true = b_values.cpu().numpy(), b_returns.cpu().numpy()
+        var_y = np.var(y_true)
+        explained_var = np.nan if var_y == 0 else 1 - np.var(y_true - y_pred) / var_y
```
Comment (yooceii) [inline cleanrl/ppo_atari_multigpu.py:384]: nit probably move under the line 387

**Comment by yooceii** [comment]
Sry, didn't have time to review before merging.

File: cleanrl/ppo_atari_multigpu.py (hunk 2 above)
Comment (vwxyzjn) [reply cleanrl/ppo_atari_multigpu.py:163]: Feel free to remove lines 89-90

File: cleanrl/ppo_atari_multigpu.py (hunk 3 above)
Comment (vwxyzjn) [reply cleanrl/ppo_atari_multigpu.py:202]: The seeding tricks done here is to ensure the same seed is used to initialize the agent's parameters: see "Adjust seed per process" https://docs.cleanrl.dev/rl-algorithms/ppo/#implementation-details_6. The more elegant way to do it is use an API to somehow broadcast the Agent's parameters in rank 0 to other tanks, but I haven't found such an API

File: cleanrl/ppo_atari_multigpu.py (hunk 4 above)
Comment (vwxyzjn) [reply cleanrl/ppo_atari_multigpu.py:222]: See comment above.

File: cleanrl/ppo_atari_multigpu.py (hunk 5 above)
Comment (vwxyzjn) [reply cleanrl/ppo_atari_multigpu.py:236]: Hmm in the past I have weird issues with `torch.tensor`. I'd also avoid changing it just in `ppo_atari_multigpu.py` but not in other files. Bottom line I don't think this would be a huge issue and cause performance differences, but I am happy to change it if evidence shows otherwise :)

File: cleanrl/ppo_atari_multigpu.py (hunk 6 above)
Comment (vwxyzjn) [reply cleanrl/ppo_atari_multigpu.py:262]: See comment above.

File: cleanrl/ppo_atari_multigpu.py (hunk 7 above)
Comment (vwxyzjn) [reply cleanrl/ppo_atari_multigpu.py:384]: This just follows the structure like in other files.

**Review by vwxyzjn — COMMENTED** [review]
Thank you @yooceii for the review! I left a few comments. Feel free to open a PR to fix applicable issues :)
